src/Routers/UsersRouter.js

```diff
@@ -289,6 +333,143 @@ export class UsersRouter extends ClassesRouter {
+        .update(String(encryptionKey))
+        .digest('base64')
+        .substr(0, 32);
+      const data = Buffer.from(mfa, 'base64');
+      const authTagLocation = data.length - 16;
+      const ivLocation = data.length - 32;
+      const authTag = data.slice(authTagLocation);
+      const iv = data.slice(ivLocation, authTagLocation);
+      const encrypted = data.slice(0, ivLocation);
+      const decipher = crypto.createDecipheriv(algorithm, encryption, iv);
+      decipher.setAuthTag(authTag);
+      return await new Promise((resolve, reject) => {
+        let decrypted = '';
+        decipher.on('readable', chunk => {
+          while (null !== (chunk = decipher.read())) {
+            decrypted += chunk.toString('utf8');
+          }
+        });
+        decipher.on('end', () => {
+          resolve(decrypted);
+        });
+        decipher.on('error', e => {
+          reject(e);
+        });
+        decipher.write(encrypted);
+        decipher.end();
+      });
+    } catch (err) {
+      throw new Parse.Error(212, 'Invalid MFA token');
+    }
+  }
+  async enableMFA(req) {
+    const mfaenabled = req.config.multiFactorAuth || {};
+    if (!mfaenabled.enabled) {
+      throw new Parse.Error(210, 'MFA is not enabled.');
+    }
+    if (!req.auth) {
+      throw new Parse.Error(101, 'Unauthorized');
+    }
+    const [user] = await req.config.database.find('_User', {
+      objectId: req.auth.user.id,
+    });
+    if (!user) {
+      throw new Parse.Error(101, 'Unauthorized');
+    }
+    if (user._mfa) {
+      throw new Parse.Error(214, 'MFA is already enabled on this account.');
+    }
+    const secret = authenticator.generateSecret();
+    const otpauth = authenticator.keyuri(user.username, req.config.appName, secret);
+    const storeKey = this.encryptMFAKey(
+      `pending:${secret}`,
+      req.config.multiFactorAuth.encryptionKey
+    );
+    await req.config.database.update('_User', { objectId: user.objectId }, { _mfa: storeKey });
+    return { response: { qrcodeURL: otpauth, secret } };
+  }
+
+  async verifyMFA(req) {
+    const mfaenabled = req.config.multiFactorAuth || {};
+    if (!mfaenabled.enabled) {
+      throw new Parse.Error(210, 'MFA is not enabled.');
+    }
+    if (!req.auth.user) {
+      throw new Parse.Error(101, 'Unauthorized');
+    }
+    const { token } = req.body;
+    if (!token) {
+      throw new Parse.Error(211, 'Please provide a token.');
+    }
+    // Fetch the user directly from the DB as we need the _mfa
+    const [user] = await req.config.database.find('_User', {
+      objectId: req.auth.user.id,
+    });
+    if (!user._mfa) {
+      throw new Parse.Error(
+        213,
+        'MFA is not enabled on this account. Please enable MFA before calling this function.'
+      );
+    }
+    const mfa = await this.decryptMFAKey(user._mfa, req.config.multiFactorAuth.encryptionKey);
+    if (mfa.indexOf('pending:') !== 0) {
+      throw new Parse.Error(214, 'MFA is already active');
```

Comment:
Requires Parse JS SDK reference for error code; check for possible inconsistency in error message for same error code.